### DISCUSSION TITLE
Validating a json based on the input values

### DISCUSSION THREAD

**Discussion by sturaw** [json-schema-org/community#229] [Q&A] [answered]
Hi, I'm someone new to json schema and while going to the docs I've found that it has a lot of functionality, most of which are based on the keys in the json. I have a very specific requirement and was wondering if that would be possible to do directly through JSON schema (found an alternative where i would have to change the structure but I'd like to avoid that extra overhead.)
In the following JSON,
```
{
    "fruit": "apple",
    "locations": [
        {
            "name": "UK",
            "value": "UH"
        },
        {
            "name": "EU",
            "value": "OH"
        },
        {
            "name": "INDIA",
            "value": "HP"
        },
        {
            "name": "USA",
            "value": "All"
        },
        {
            "name": "USA1",
            "value": "ABC"
        },
        {
            "name": "USA2",
            "value": "ABC2"
        }

    ]
}
```
My requirement is that

1. ln 'locations', 'USA' should be a mandatory in input.
2. If the value of 'USA' is 'ALL' then, 'USA1' and 'USA2' should be mandatory.

Would appreciate any help.

**Comment by gregsdennis** [comment, marked as answer]
This can be done fairly easily if you're using a draft 7 or later.

What you want is going to be a combination of [this StackOverflow answer](https://stackoverflow.com/a/73457576/878701) and the `contains` keyword.

### For (1)

You want at least one item to have `name: USA`.

`contains` alone solves this for you.  `contains` defines a schema, and if the value is an array, it must contain at least one item that matches that schema.  It doesn't matter where in the array the item is.

```
{
  ...
  "contains": {
    "type": "object",
    "properties": {
      "name": { "const": "USA" }
    },
    "required": [ "name" ]
  }
  ...
}
```

*If you've already required that all of the items are objects with `name` and `value`, you can leave out the `type` and `required` from contains, but it doesn't hurt anything to overconstrain like this.*

### For (2)

You want a conditional that _if_ the item with `name: USA` also has `value: All`, then your array must also contain an item with `name: USA1` and another item with `name: USA2`.

First, for the conditional, you want to use `if`/`then` (`else` is also available, but you don't need it here).

`if` and `then` each define schemas.  If the value passes the `if` schema, then it must also pass the `then` schema.

So for `if`, we check to see if there is an item with `name: USA, value: All`.  This is very similar to what we have above.

```
{
  "if": {
    "contains": {
      "type": "object",
      "properties": {
        "name": { "const": "USA" },
        "value": { "const": "All" }
      },
      "required": [ "name", "value" ]
    }
  }
}
```

And for the `then` we check to see if both of the other items are there.  To combine two `contains` keywords, we can use an `allOf`.

```
{
  "then": {
    "allOf": [
      {
        "contains": {
          "type": "object",
          "properties": {
            "name": { "const": "USA1" }
          },
          "required": [ "name" ]
        }
      },
      {
        "contains": {
          "type": "object",
          "properties": {
            "name": { "const": "USA2" }
          },
          "required": [ "name" ]
        }
      }
    ]
  }
}
```

Putting all of this together, you get:

```
{
  ...
  "contains": {
    "type": "object",
    "properties": {
      "name": { "const": "USA" }
    },
    "required": [ "name" ]
  },
  "if": {
    "contains": {
      "type": "object",
      "properties": {
        "name": { "const": "USA" },
        "value": { "const": "All" }
      },
      "required": [ "name", "value" ]
    }
  },
  "then": {
    "allOf": [
      {
        "contains": {
          "type": "object",
          "properties": {
            "name": { "const": "USA1" }
          },
          "required": [ "name" ]
        }
      },
      {
        "contains": {
          "type": "object",
          "properties": {
            "name": { "const": "USA2" }
          },
          "required": [ "name" ]
        }
      }
    ]
  }
  ...
}
```

**Reply by handrews** [reply]
I'll just add to Greg's excellent answer that if you're using draft-06 or earlier, you can still do this, but you have to use different keywords.  Using `{A}`, `{B}`, `{C}` as placeholder schemas because I'm lazy, this draft-07+ structure:

```JSON
{
  "if": {A},
  "then": {B},
  "else": {C}
}
```
becomes this in draft-06 and earlier:
```JSON
{
  "oneOf": [
    "allOf": [{A}, {B}],
    "allOf": [{"not": {A}}, {C}]
  ]
}
```

And if you're using draft-04 (which hopefully you aren't), then you have to replace `"const": "foo"` with `"enum": ["foo"]`
